README.md
# Collapsible Sliding Panel

## Overview

The Collapsible Sliding Panel component used to show/hide contained elements.

[CHANGELOG.md](CHANGELOG.md) contains Release Notes for this package.

[CONTRIBUTING.md](CONTRIBUTING.md) contains guidelines on how to help improve this package.

## Using Collapsible Sliding Panel

The Collapsible Sliding Panel can be included as follows:

```jsx
import {CollapsibleSlidingPanel} from 'collapsible-sliding-panel';

...

getSlidingPanelContent() {
  [some logic to build sliding panel content]
}

getMainPanelContent() {
  [some logic to build main panel content]
}

...

render () {
    ....
    return (
        <CollapsibleSlidingPanel
          slidingPanelClassName='collapsible-sliding-panel-classname'
          slidingPanelClosedClassName='collapsible-sliding-panel-is-closed-classname'
          expanderHandleClassName='collapsible-sliding-panel-expander-classname'
          slidingPanelContent={getSlidingPanelContent()}
          collapsiblePanelOpen={true}
          tooltipText={{
            expand: "Expandir",
            collapse: "Contraer"
          }}>
            {getMainPanelContent()}
        </CollapsibleSlidingPanel>
    )
}
```

## Configuring Collapsible Sliding Panel

The Collapsible Sliding Panel can be configured as follows

Property Name | Description
--------------|------------
slidingPanelClassName | The class name used for styling the sliding panel
slidingPanelClosedClassName | The class name used for styling the sliding panel when it is closed
expanderHandleClassName | The class name used for styling the expander handle
slidingPanelContent | The content that is to be rendered inside the sliding panel
collapseCallback | The callback function that is called when the sliding panel is expanded or collapsed
collapsiblePanelOpen | The boolean sets the initial expanded (true) / collapsed (false) state of the sliding panel; defaults to collapsed (false)
tooltipText | Values appear as tooltips on the panel button: `tooltipText.expand` and `tooltipText.collapse`. `tooltipText` is optional and defaults to "Expand" and "Collapse", respectively.

Style class names were added as properties for this component so that different styles could be assigned to different instances of the sliding panel within a single application.

When using the default styles, the 3 style class names mentioned above are combined in order to achieve the 'sliding/collapsing' effect as follows:

.slidingPanelClassName - how the sliding panel appears when it is expanded ... it also uses the CSS 'transition' property to determine it's expanding functionality
.slidingPanelClassName.slidingPanelClosedClassName - how the sliding panel appears when it is closed ... it also uses the CSS 'transition' property to determine it's collapsing functionality
.expanderHandleClassName - how the expander handle appears ... it also uses the CSS 'transition' property to determine how it moves when the sliding panel opens
.expanderHandleClassName.slidingPanelClosedClassName - how the expander handle appears when it is closed ... it also uses the CSS 'transition' property to determine how it moves when the sliding panel closes

## Run & Test Locally

The Collapsible Sliding Panel component can be launched independently in order to see/test it's capabilities as follows:

1. type the following NPM command from a comand prompt that supports NPM: npm run startDev
2. view the test app at the following URL: <http://localhost:8005/collapsible-sliding-panel/>

Once launched, the test application consists of the following components:

- header bar ... application title and text indicating whether the sliding panel is opened or closed
- sliding panel ... initially this panel is collapsed and it's content is not visible
- main content panel ... this is where the Collapsible Sliding Panel will render it's child elements (this.props.children)
- expander handle ... this 'tab' (depending on style assigned) is used to trigger the expanding and collapsing of the sliding panel

For the test application and using default styles, the Collapsible Sliding Panel looks like ths when collapsed:
![Image of Collapsible Sliding Panel when panel is collapsed](screen_cap_sliding_panel_closed.JPG)

For the test application and using default styles, the Collapsible Sliding Panel looks like ths when expanded:
![Image of Collapsible Sliding Panel when panel is expanded](screen_cap_sliding_panel_open.JPG)


CHANGELOG.md
## 2.3.0 (September 15, 2020)
- Add ability to override tooltip text for component

## 2.2.0 (August 11, 2020)
- Optional prop to set whether the collapsible panel is initially open or closed (default is closed)

## 2.1.0 (June 15, 2020)

- bind collapse handler callback to `this`, so method can call `this.setState`
- add `CHANGELOG.md`
- add `CONTRIBUTING.md`

## 0.0.1

- Initial component creation


webpack.devConfig.js
/*
 * Copyright © 2016-2017 European Support Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Created by KACHUNA on 10/21/2016.
 */
var webpack = require('webpack');
var CopyWebpackPlugin = require('copy-webpack-plugin');
var path = require('path');
var devPort = process.env.PORT || 8005;

var PATHS = {
  SRC: path.resolve(__dirname, '.'),
  TARGET: path.resolve(__dirname, 'dist')
};

var devConfig = {
  entry: {
    bundle: [
      './main.collapsibleSlidingPanel.jsx',
      `webpack-dev-server/client?http://localhost:${devPort}`,
      'webpack/hot/only-dev-server'
    ]
  },
  output: {
    path: path.join(__dirname, 'dist'),
    publicPath: `http://localhost:${devPort}/collapsible-sliding-panel/`,
    filename: '[name].js'
  },
  devServer: {
    port: devPort,
    historyApiFallback: true,
    publicPath: `http://localhost:${devPort}/collapsible-sliding-panel/`,
    contentBase: path.join(__dirname, 'dist'),
    hot: true,
    progress: true,
    inline: true,
    debug: true,
    stats: {
      colors: true
    }
  },
  module: {
    preLoaders: [{
      test: /\.(js|jsx)?$/,
      loader: 'eslint-loader',
      exclude: /node_modules/
    }],
    loaders: [
      {test: /\.(js|jsx)$/, loaders: ['react-hot', 'babel-loader', 'eslint-loader'], exclude: /node_modules/},
      {test: /\.(png|jpg|svg)(\?.*)?$/, loader: 'url-loader?limit=16384'},
      {test: /\.(css|scss)$/, loaders: ['style', 'css?sourceMap', 'sass?sourceMap']}
    ]
  },
  plugins: [
    new CopyWebpackPlugin([
      {
        from: path.resolve(PATHS.SRC, 'index.html'), to: 'index.html', force: true
      }
    ]),
    new webpack.DefinePlugin({
      DEBUG: true
    }),

    new webpack.HotModuleReplacementPlugin()
  ]
};

module.exports = devConfig;

webpack.dev.config.js
// for Hot Module Replacement
const path = require('path');
const webpack = require('webpack');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const merge = require('webpack-merge');
const core = require('./webpack.core.config');

module.exports = merge(core, {
  mode: "development",
  devtool: 'eval-source-map',
  entry: {
    'vertical-filter-bar' : './dev/src/index.js'
  },
  output: {
    filename: '[name].bundle.js',
    path: path.resolve(__dirname, "dist")
  },
  module: {
    rules: [
      {
        test: /\.(css|sass|scss)$/,
        use: [
          "style-loader", //3. Inject styles into DOM
          "css-loader", //2. Turns css into commonjs
          "sass-loader" //1. Turns sass into css
        ]
      }
    ]
  },
  plugins: [
    // use HtmlWebpackPlugin to auto generate the index.html file
    // based on a template (index_template)
    new HtmlWebpackPlugin({
      title: "Collapsible sliding panel - Demo",
      template: './index_template.html',
      filename: 'index.html'
    }),
    new webpack.HotModuleReplacementPlugin()
  ],
  devServer: {
    // Silence WebpackDevServer's own logs since they're generally not useful.
    // It will still show compile warnings and errors with this setting.
    clientLogLevel: 'none',
    // Enable gzip compression of generated files.
    compress: true,
    // where to pull the source from
    contentBase: '/dist',
    // enable hot module replacement
    hot: true,
    // launch a browser by default on start up
    open: true
  }
});

webpack.core.config.js
// Core/base webpack config (to be combined with prod or dev config
const CircularDependencyPlugin = require('circular-dependency-plugin')

module.exports = {
  resolve: {
    extensions: ['*', '.js', '.jsx']
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        use: ["source-map-loader"],
        enforce: "pre"
      },
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: ['@babel/preset-env', '@babel/preset-react'],
              plugins: ['@babel/plugin-proposal-class-properties']
            }
          },
          'eslint-loader'
        ]
      },
      {
        test: /\.(jpe?g|png|gif|svg)$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[name].[ext]',
              outputPath: 'dist/images'
            }
          }
        ],
        exclude: /node_modules/,
      }
    ]
  },
  plugins: [
    new CircularDependencyPlugin({
      // exclude detection of files based on a RegExp
      exclude: /a\.js|node_modules/,
      // add errors to webpack instead of warnings
      failOnError: true,
      // set the current working directory for displaying module paths
      cwd: process.cwd(),
    })
  ]
};

webpack.prod.config.js
const path = require('path');
const merge = require('webpack-merge');
const core = require('./webpack.core.config');
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const TerserJSPlugin = require('terser-webpack-plugin');
const OptimizeCSSAssetsPlugin = require('optimize-css-assets-webpack-plugin');

module.exports = merge(core, {
  mode: 'production',
  devtool: 'source-map',
  entry: {
    'collapsible-sliding-panel' : './index.js'
  },
  output: {
    path:  path.join(__dirname, 'dist'),
    filename: '[name].bundle.js',
    publicPath: '/',
    library: 'collapsibleSlidingPanel',
    libraryTarget: 'umd'
  },
  externals: [
    'react-dom',
    'react',
    'prop-types'
  ],
  module: {
    rules: [
      {
        test: /\.(css|sass|scss)$/,
        use: [
          "style-loader", //3. Inject styles into DOM
          "css-loader", //2. Turns css into commonjs
          "sass-loader" //1. Turns sass into css
        ]
      }
    ]
  },
  plugins: [
    new CleanWebpackPlugin()
  ],
  optimization: {
    minimizer: [
      new TerserJSPlugin({ sourceMap: true }),
      new OptimizeCSSAssetsPlugin({})
    ]
  }
});

webpack.config.js
/*
 * Copyright © 2016-2017 European Support Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var webpack = require('webpack');
var path = require('path');
var devPort = process.env.PORT || 8005;

var PATHS = {
  SRC: path.resolve(__dirname, '.'),
  TARGET: path.resolve(__dirname, 'dist')
};

var config = {
  entry: [path.resolve(PATHS.SRC, 'index.js')],
  output: {
    path: PATHS.TARGET,
    filename: 'bundle.js',
    library: 'collapsible-sliding-panel',
    libraryTarget: 'umd'
  },
  resolve: {
    root: [path.resolve('.')]
  },
  eslint: {
    failOnWarning: false,
    failOnError: true,
    configFile: '.eslintrc'
  },
  devtool: 'source-map',
  module: {
    preLoaders: [{
      test: /\.(js|jsx)?$/,
      loader: 'eslint-loader',
      exclude: /node_modules/
    }],
    loaders: [
      {test: /\.(js|jsx)$/, loaders: ['react-hot', 'babel-loader', 'eslint-loader'], exclude: /node_modules/},
      {test: /\.(png|jpg|svg)(\?.*)?$/, loader: 'url-loader?limit=16384'},
      {test: /\.(css|scss)$/, loaders: ['style', 'css?sourceMap', 'sass?sourceMap']}
    ]
  },
  devServer: {
    port: devPort,
    historyApiFallback: true,
    publicPath: `http://localhost:${devPort}/collapsible-sliding-panel/`,
    contentBase: PATHS.TARGET,
    hot: true,
    progress: true,
    inline: true,
    debug: true,
    stats: {
      colors: true
    }
  }
};

module.exports = config;